### ISSUE TITLE
rstudio: issues with darwin frameworks

### ISSUE THREAD

**Issue by jbaum98** [NixOS/nixpkgs#24237]
I'm working on updating R-studio and adding Darwin support. I'm having an issue where the compiler can't find `'CoreFoundation/CFAttributedString.h'` even though build dependencies include`CoreServices`, `AppKit`, and `ApplicationServices`. I added `darwin.apple_sdk.sdk` to the build dependencies, and now it can't find `'AE/AE.h'`. Should I be adding different build dependencies, or is this some issue related to pure/impure CoreFoundation as related to #23946 and #22571?

**Comment by domenkozar** [comment]
cc @LnL7 @copumpkin

**Comment by copumpkin** [comment]
I think AE.framework only appeared in the 10.11 SDK, which we haven't packaged yet because of some trickiness around CoreFoundation. How essential is the AE support to rstudio? Can it be built without it for now?

**Comment by jbaum98** [comment]
I found AE.fraemwork in the 10.9 SDK, buried within CoreServices. Is there
a reason why it isn't being seen?
/nix/store/ydbyzk39rcvywdmb67d2vcv0cnl8ab1c-MacOS_SDK-10.9/Library/Frameworks/CoreServices.framework/Versions/Current/Frameworks/AE.framework

**Comment by copumpkin** [comment]
Oh, interesting. Not sure! It's probably in 10.10 then, too. Have you tried adding `apple_sdk.frameworks.CoreServices` explicitly to the `buildInputs`?

**Comment by copumpkin** [comment]
We might be screwing up some of the subframework stuff.

**Comment by jbaum98** [comment]
Yeah I took away the sdk and just added CoreServices. That didn't work because it isn't looking at subframeworks, but when I added manually the subframework Framework path to `NIX_CFLAGS_COMPILE` that fixed it. I had to do that also for ApplicationServices, and now its compiling.
Now I have a different linker issue where it can't find some symbols:
```
Undefined symbols for architecture x86_64:
  "_OBJC_CLASS_$_NSArray", referenced from:
      objc-class-ref in AppDelegate.mm.o
      objc-class-ref in GwtCallbacks.mm.o
      objc-class-ref in Options.mm.o
      objc-class-ref in SecondaryWindowController.mm.o
      objc-class-ref in WebViewController.mm.o
  "_OBJC_CLASS_$_NSData", referenced from:
      objc-class-ref in Utils.mm.o
  "_OBJC_CLASS_$_NSDictionary", referenced from:
      objc-class-ref in DockTileView.mm.o
      objc-class-ref in GwtCallbacks.mm.o
      objc-class-ref in MainFrameMenu.mm.o
      objc-class-ref in Options.mm.o
  "_OBJC_CLASS_$_NSException", referenced from:
      objc-class-ref in WebViewController.mm.o
  "_OBJC_CLASS_$_NSLocale", referenced from:
      objc-class-ref in Utils.mm.o
  "_OBJC_CLASS_$_NSMutableArray", referenced from:
      objc-class-ref in MainFrameMenu.mm.o
      objc-class-ref in Options.mm.o
  "_OBJC_CLASS_$_NSMutableDictionary", referenced from:
      objc-class-ref in WebViewController.mm.o
  "_OBJC_CLASS_$_NSTimer", referenced from:
      objc-class-ref in AppDelegate.mm.o
  "_OBJC_CLASS_$_NSURL", referenced from:
      objc-class-ref in GwtCallbacks.mm.o
      objc-class-ref in SessionLauncher.mm.o
      objc-class-ref in WebViewController.mm.o
  "_OBJC_CLASS_$_NSUserDefaults", referenced from:
      objc-class-ref in Options.mm.o
      objc-class-ref in Utils.mm.o
  "_OBJC_EHTYPE_$_NSException", referenced from:
      GCC_except_table6 in GwtCallbacks.mm.o
      GCC_except_table9 in GwtCallbacks.mm.o
      GCC_except_table46 in WebViewController.mm.o
ld: symbol(s) not found for architecture x86_64
```
Not sure if this is related.

**Comment by copumpkin** [comment]
I think we'll probably need to add `CoreFoundation` to pkgs/os-specific/darwin/apple-sdk/frameworks.nix and have you depend on that. I think the issue is that the CF we bundle by default isn't the full Apple one because they don't release full source for it (or didn't until recently, and now it's all full of swift-specific stuff which also isn't compatible with the system one)

**Comment by jbaum98** [comment]
One other thing I did was I depended on `darwin.cf-private`, what's the difference between that and `CoreFoundation`?

**Comment by jbaum98** [comment]
I found the missing symbols in an object file located in `Foundation` using `nm`. Not sure how to include them.

**Comment by jbaum98** [comment]
Just kidding, I can see now that those symbols are dynamically linked in `Foundation` to `/System/Library/Frameworks/CoreFoundation.framework/Version
s/A/CoreFoundation`. Adding that to the linker makes it compile. So it is a CoreFoundation issue.

**Comment by jbaum98** [comment]
So what should I do to fix it? Right now I'm adding `-F/System/Library/Frameworks` to `NIX_LDFLAGS` and its compiling, but that's obviously not ideal.

**Comment by TravisWhitaker** [comment]
@jbaum98 I'm having the same issue building a Haskell program. Despite adding CoreFoundation to `buildInputs` in the Nix expression and `frameworks` in the cabal file, I can only link the executable if I add `-F/System/Library/Frameworks`. @copumpkin has any progress been made here? Perhaps there's a package under `darwin` somewhere that at least captures this impurity?

**Comment by TravisWhitaker** [comment]
Hmm, it seems that `darwin.cf-private` _should_ do what we need here, but either the `shellHook` isn't being run for some reason, or something else is stomping on `$NIX_LDFAGS`.
